x/protorev/keeper/rebalance.go

```diff
@@ -167,30 +171,174 @@ func (k Keeper) FindMaxProfitForRoute(ctx sdk.Context, route RouteMetaData, rema
+	return k.ExtendSearchRangeIfNeeded(ctx, route, inputDenom, curLeft, curRight, updatedMax)
+}
+
+// CalculateUpperBoundForSearch returns the max amount in that can be used for the binary search
+// respecting the max ticks moved across all concentrated liquidity pools in the route.
+func (k Keeper) CalculateUpperBoundForSearch(
+	ctx sdk.Context,
+	route RouteMetaData,
+	inputDenom string,
+) (sdk.Int, error) {
+	var intermidiateCoin sdk.Coin
+
+	// Iterate through all CL pools and determine the maximal amount of input that can be used
+	// respecting the max ticks moved.
+	for index := route.Route.Length() - 1; index >= 0; index-- {
+		hop := route.Route[index]
+		pool, err := k.poolmanagerKeeper.GetPool(ctx, hop.PoolId)
+		if err != nil {
+			return sdk.ZeroInt(), err
+		}
+
+		tokenInDenom := inputDenom
+		if index > 0 {
+			tokenInDenom = route.Route[index-1].TokenOutDenom
+		}
+
+		switch {
+		case pool.GetType() == poolmanagertypes.Concentrated:
+			// If the pool is a concentrated liquidity pool, then check the maximum amount in that can be used
+			// and determine what this amount is as an input at the previous pool (working all the way up to the
+			// beginning of the route).
+			maxTokenIn, maxTokenOut, err := k.concentratedLiquidityKeeper.ComputeMaxInAmtGivenMaxTicksCrossed(
+				ctx,
+				pool.GetId(),
+				tokenInDenom,
+				types.MaxTicksCrossed,
+			)
+			if err != nil {
+				return sdk.ZeroInt(), err
+			}
+
+			// if there have been no other CL pools in the route, then we can set the intermediate coin to the max input amount.
+			// Additionally, if the amount of the previous token is greater than the possible amount from this pool, then we
+			// can set the intermediate coin to the max input amount (from the current pool). Otherwise we have to do a
+			// safe swap given the previous max amount.
+			if intermidiateCoin.IsNil() || maxTokenOut.Amount.LT(intermidiateCoin.Amount) {
+				intermidiateCoin = maxTokenIn
+				continue
+			}
+
+			// In the scenario where there are multiple CL pools in a route, we select the one that inputs
+			// the smaller amount to ensure we do not overstep the max ticks moved.
+			intermidiateCoin, err = k.executeSafeSwap(ctx, pool.GetId(), intermidiateCoin, tokenInDenom, route.StepSize)
+			if err != nil {
+				return sdk.ZeroInt(), err
+			}
+		case !intermidiateCoin.IsNil():
+			// If we have already seen a CL pool in the route, then simply propagate the intermediate coin up
+			// the route.
+			intermidiateCoin, err = k.executeSafeSwap(ctx, pool.GetId(), intermidiateCoin, tokenInDenom, route.StepSize)
+			if err != nil {
+				return sdk.ZeroInt(), err
+			}
+		}
+	}
+
+	// In the case where there are no CL pools, we want to return the extended max input amount
+	if intermidiateCoin.IsNil() {
+		return types.ExtendedMaxInputAmount, nil
+	}
+
+	// Ensure that the normalized upper bound is not greater than the extended max input amount
+	upperBound := intermidiateCoin.Amount.Quo(route.StepSize)
```

Comment:
Is it possible for a route's step size to be zero?

Reply:
tldr no, we manually set these and enforce that the step size cannot be zero (when we run validate basic on a msg).